modyn/evaluator/internal/pytorch_evaluator.py

```diff
@@ -0,0 +1,177 @@
+import io
+import logging
+import multiprocessing as mp
+import os
+import pathlib
+import queue
+import traceback
+from inspect import isfunction
+from typing import Any, Callable, Optional, Union
+
+import torch
+from modyn.evaluator.internal.dataset.evaluation_dataset import EvaluationDataset
+from modyn.evaluator.internal.utils import EvaluationInfo, EvaluatorMessages
+
+EVALUATION_TRANSFORMER_FUNC_NAME = "evaluation_transformer"
+LABEL_TRANSFORMER_FUNC_NAME = "label_transformer_function"
+
+
+class PytorchEvaluator:
+    def __init__(
+        self,
+        evaluation_info: EvaluationInfo,
+        status_query_queue: mp.Queue,
+        status_response_queue: mp.Queue,
+        logger: logging.Logger,
+    ) -> None:
+        self.logger = logger
+        self._evaluation_id = evaluation_info.evaluation_id
+
+        self._model = evaluation_info.model_handler(
+            evaluation_info.model_configuration_dict, evaluation_info.device, evaluation_info.amp
+        )
+        self._load_state(evaluation_info.model_path)
+
+        # setup dataloaders
+        self._info("Setting up data loaders.")
+        self._dataloader = self._prepare_dataloader(evaluation_info)
+
+        self._mod_dict: dict[str, Any] = {}
+        self._evaluation_transformer = self._setup_transformer_function(
+            evaluation_info.evaluation_transformer, EVALUATION_TRANSFORMER_FUNC_NAME
+        )
+
+        self._metrics = evaluation_info.metrics
+
+        self._label_tranformer_function = self._setup_transformer_function(
+            evaluation_info.label_transformer, LABEL_TRANSFORMER_FUNC_NAME
+        )
+
+        self._device = evaluation_info.device
+        self._device_type = "cuda" if "cuda" in self._device else "cpu"
+        self._amp = evaluation_info.amp
+
+        self._status_query_queue = status_query_queue
+        self._status_response_queue = status_response_queue
+
+        self._num_samples = 0
+
+        self._info("Initialized PyTorch evaluator.")
+
+    def _prepare_dataloader(self, evaluation_info: EvaluationInfo) -> torch.utils.data.DataLoader:
+        self._debug("Creating EvaluationDataset.")
+        dataset = EvaluationDataset(
+            evaluation_info.dataset_id,
+            evaluation_info.bytes_parser,
+            evaluation_info.transform_list,
+            evaluation_info.storage_address,
+            evaluation_info.evaluation_id,
+        )
+        self._debug("Creating DataLoader.")
+        dataloader = torch.utils.data.DataLoader(
+            dataset, batch_size=evaluation_info.batch_size, num_workers=evaluation_info.num_dataloaders
+        )
+
+        return dataloader
+
+    def _setup_transformer_function(self, serialized_code: str, func_name: str) -> Optional[Callable]:
+        if serialized_code == "":
+            return None
```

Comment:
why is that required? we did not have that in the trainer server, right?

Reply:
It is required as we want to differentiate between empty string = no function and a malformed serialized function where we throw an error. This was also done similarly in the trainer server. But maybe I'm missing something here.

Reply:
Ah, ok, I'm not aware that this was a problem in the trainer server, but ok makes sense